go/chat/boxer.go

```diff
@@ -402,8 +412,276 @@ func (b *Boxer) unboxV1(ctx context.Context, boxed chat1.MessageBoxed, encryptio
+	}, nil
+}
+
+// Unversions a header.
+// Also check that the HeaderSignature field from MessageBoxed V1 is nil.
+// Therefore only for use with MessageBoxed V2.
+// Returns (header, bodyHash, err)
+func (b *Boxer) unversionHeader(ctx context.Context, headerVersioned chat1.HeaderPlaintext) (chat1.MessageClientHeaderVerified, []byte, UnboxingError) {
+	headerVersion, err := headerVersioned.Version()
+	if err != nil {
+		return chat1.MessageClientHeaderVerified{}, nil, NewPermanentUnboxingError(err)
+	}
+	switch headerVersion {
+	case chat1.HeaderPlaintextVersion_V1:
+		hp := headerVersioned.V1()
+		if hp.HeaderSignature != nil {
+			return chat1.MessageClientHeaderVerified{}, nil,
+				NewPermanentUnboxingError(fmt.Errorf("HeaderSignature non-nil in MBV2"))
+		}
+		return chat1.MessageClientHeaderVerified{
+			Conv:         hp.Conv,
+			TlfName:      hp.TlfName,
+			TlfPublic:    hp.TlfPublic,
+			MessageType:  hp.MessageType,
+			Prev:         hp.Prev,
+			Sender:       hp.Sender,
+			SenderDevice: hp.SenderDevice,
+			// CORE-4540: MerkleRoot will be in signed header.
+			OutboxID:   hp.OutboxID,
+			OutboxInfo: hp.OutboxInfo,
+		}, hp.BodyHash, nil
+	default:
+		return chat1.MessageClientHeaderVerified{}, nil,
+			NewPermanentUnboxingError(NewHeaderVersionError(headerVersion,
+				b.headerUnsupported(ctx, headerVersion, headerVersioned)))
+	}
+}
+
+func (b *Boxer) unversionBody(ctx context.Context, bodyVersioned chat1.BodyPlaintext) (chat1.MessageBody, UnboxingError) {
+	bodyVersion, err := bodyVersioned.Version()
+	if err != nil {
+		return chat1.MessageBody{}, NewPermanentUnboxingError(err)
+	}
+	switch bodyVersion {
+	case chat1.BodyPlaintextVersion_V1:
+		return bodyVersioned.V1().MessageBody, nil
+	default:
+		return chat1.MessageBody{},
+			NewPermanentUnboxingError(NewBodyVersionError(bodyVersion,
+				b.bodyUnsupported(ctx, bodyVersion, bodyVersioned)))
+	}
+}
+
+func (b *Boxer) verifyBodyHash(ctx context.Context, bodyEncrypted chat1.EncryptedData, bodyHashSigned []byte) UnboxingError {
+	bodyHashObserved := b.hashV1(bodyEncrypted.E)
+
+	if len(bodyHashSigned) == 0 {
+		return NewPermanentUnboxingError(BodyHashInvalid{})
+	}
+
+	if !libkb.SecureByteArrayEq(bodyHashObserved, bodyHashSigned) {
+		return NewPermanentUnboxingError(BodyHashInvalid{})
+	}
+	return nil
+}
+
+// Compare the unsigned and signed header for MessageBoxedVersion_V1.
+// The V1 and V2 checks are different methods because they are strict on slightly different things.
+// Confirm that fields in the server-supplied ClientHeader match what
+// we decrypt. It would be preferable if the server didn't supply this data
+// at all (so that we didn't have to worry about anyone trusting it
+// *before* we get to this check, for example), but since we have it we
+// need to check it.
+// The most important check here is that the Sender and SenderDevice match.
+// That is the only thing that gives the verification key used credibility.
```

Comment:
Confirming my understanding of the tricky details here: It would _almost_ be ok for us to ignore the Sender entirely here, and instead look everything up via the signing KID. However, that opens us up to a weird attack where someone delegates a crazy weak sibkey that can return true for any signature, and then with evil Keybase's cooperation they claim authorship of other people's messages in an existing convo (that they were already a writer in). Preventing that requires that the message asserts its own sender, so that fixing the hash of the message via prev pointers fixes the sender, and that means we have to check it for consistency.

Followup thought: We thought about having signencrypt do this for us. The simplest possible solution, where we hash the signing key in with all the other associated data, doesn't actually work, because the weak signing key would still return true for the mismatching signature input. If we transport the public key alongside the plaintext (increasing the size of the final ciphertext, unlike the first idea), that could work, but the receiving client would have to remember to compare the keys. Another option would be to have the _encryption_ layer take the signing key as associated data, which clients can't forget to check. (NaCl's crypto_secretbox doesn't support associated data directly, but the standard way to do it is to hash the encryption key together with your AD, and use the result to encrypt the actual message, so that decryption breaks if the AD changes.)